### PR TITLE
feat(condo): DOMA-5776 added ticket feedback by resident (deprecated review)

### PR DESCRIPTION
**Pull request by Alllex202**
<s>based on https://github.com/open-condo-software/condo/pull/3134</s>

`reviewValue` and `reviewComment` will be remove when mobile apps stopped use them.

Снимок экрана 2023-04-14 в 20 12 15

### CODE REVIEW

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 1]
```diff
@@ -59,6 +67,24 @@ const { RESIDENT } = require('@condo/domains/user/constants/common')
...
+        code: BAD_USER_INPUT,
+        type: WRONG_VALUE,
+        message: '"feedbackValue" must be specified if there is "feedbackAdditionalOptions" or "feedbackComment"',
+        messageForUser: 'api.ticket.FEEDBACK_VALUE_MUST_BE_SPECIFIED',
+    },
+}
+
+// todo(doma-5776): should delete this function when will done task DOMA-5319
+const convertTextOptionToKeyOption = (option) => {
+    if (option === 'Сделали на совесть' || option === 'Made in good faith') return FEEDBACK_ADDITIONAL_OPTIONS_BY_KEY.HIGH_QUALITY
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:81]: Hardcode! You shold get some const value inside API. And do this convert on the front-end side

**Review by nomerdvadcatpyat — APPROVED** [review]

File: apps/condo/domains/ticket/components/TicketId/TicketFeedbackFields.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    const optionsMessages: OptionsMessagesType = useMemo(() => ({
+        [FeedbackAdditionalOptionsType.LowQuality]: LowQualityMessage.toLowerCase(),
+        [FeedbackAdditionalOptionsType.Slowly]: SlowlyMessage.toLowerCase(),
+        [FeedbackAdditionalOptionsType.HighQuality]: HighQualityMessage.toLowerCase(),
+        [FeedbackAdditionalOptionsType.Quickly]: QuicklyMessage.toLowerCase(),
+    }), [HighQualityMessage, LowQualityMessage, QuicklyMessage, SlowlyMessage])
+
+    const renderOptions = useMemo(() => {
+        if (isEmpty(ticket.feedbackAdditionalOptions) || !ticket.feedbackValue) return null
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/components/TicketId/TicketFeedbackFields.tsx:51]: Yes, only `feedbackValue` can be passed.
Actually this component should not render if there is no `feedbackValue`, but I handled this case too.

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 3]
```diff
@@ -472,13 +522,58 @@ const Ticket = new GQLListSchema('Ticket', {
                 resolvedData.status = STATUS_IDS.OPEN
             }
 
+            // todo(doma-5833):  delete this block when the mobile app will use 'feedback*' fields
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:525]:
```suggestion
            // TODO(DOMA-5833):  delete this block when the mobile app will use 'feedback*' fields
```

File: apps/condo/domains/ticket/schema/Ticket.test.js [hunk 4]
```diff
@@ -1716,16 +1835,63 @@ describe('Ticket', () => {
                     status: { connect: { id: STATUS_IDS.COMPLETED } },
                 })
 
+                console.log('ticket', ticket)
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.test.js:1838]: Debug?

File: apps/condo/domains/ticket/schema/Ticket.test.js [hunk 5]
```diff
@@ -1716,16 +1835,63 @@ describe('Ticket', () => {
...
+
                 const [updatedTicket] = await updateTestTicket(userClient, ticket.id, {
                     reviewValue: REVIEW_VALUES.RETURN,
                 })
+                console.log('updatedTicket', updatedTicket)
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.test.js:1843]:
```suggestion

```

File: apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+--
+-- [CUSTOM] Add some helper functions
+--
+
+CREATE OR REPLACE FUNCTION getFeedbackOptionKeyFromReviewCommentOption(option text)
+    returns text
+as
+$getFeedbackOptionKeyFromReviewCommentOption$
+BEGIN
+    if option = 'Сделали на совесть' or option = 'Made in good faith' then return 'highQuality';
```
Comment (pahaz) [inline apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js:86]: You can select more than one options

File: apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+CREATE OR REPLACE FUNCTION convertReviewCommentToFeedbackAdditionalOptions(value text)
+    returns jsonb
+as
+$convertReviewCommentToFeedbackAdditionalOptions$
+declare
+    key1 text;
+    key2 text;
+BEGIN
+    key1 := getFeedbackOptionKeyFromReviewCommentOption(split_part(value, ';', 1));
```
Comment (pahaz) [inline apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js:104]: You also need to trim spaces

File: apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    key2 := getFeedbackOptionKeyFromReviewCommentOption(split_part(value, ';', 2));
```
Comment (pahaz) [inline apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js:105]: What if we have more than one options?

File: apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+UPDATE public."TicketChange"
+    SET "feedbackAdditionalOptionsTo" = convertReviewCommentToFeedbackAdditionalOptions("reviewCommentTo")
+    WHERE "reviewCommentTo" IS NOT NULL AND "reviewCommentTo" != '';
+
+--
+-- [CUSTOM] Set feedbackValue from reviewValue for TicketFilterTemplate
+--
+
+UPDATE public."TicketFilterTemplate"
```
Comment (pahaz) [inline apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js:169]: You also need to check the front end side. If you use the old front end version. If should not get any runtime errors

File: apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js (hunk 8 above)
Comment (Alllex202) [reply apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js:105]: We only have reviewComment in the format `{option1}` and `{option1};{option2}`. Also with `null` and ` ` (empty line) values.

All these cases are handled

Comment (Alllex202) [reply apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js:105]: On prod there are only correct data, in right format

File: apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js (hunk 7 above)
Comment (Alllex202) [reply apps/condo/migrations/20230419122709-0261_ticket_feedbackadditionaloptions_and_more.js]: added trim spaces, but I don't think it's necessary. After all, we use a well-known data format for us

File: apps/condo/domains/ticket/access/Ticket.js [hunk 10]
```diff
@@ -88,6 +92,12 @@ async function canManageTickets ({ authentication: { item: user }, operation, it
             return ticket.client === user.id
         }
     } else {
+        // NOTE: service user with any fields pass
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/Ticket.js:95]: probably you need to check it for service user too.

Comment (Alllex202) [reply apps/condo/domains/ticket/access/Ticket.js]: fixed

**Review by pahaz — APPROVED** [review]

**Pull request merged**